### DISCUSSION TITLE
how to deal with errors?

### DISCUSSION THREAD

**Discussion by maia** [RhetTbull/osxphotos#1443] [Q&A]
I'm using osxphotos to backup my photos to a NAS (which is then synced to the cloud), using the command below. There's a few issues I don't know how to solve though and hope someone can provide a solution:

The report of the last run is: `Processed: 16926 photos, exported: 1, updated: 16925, skipped: 0, updated EXIF data: 16926, missing: 24, error: 28, touched date: 16926` (seems my NAS disconnected 1 photo before completion on the first run).

1. Why is the album for the missing photos in the Photos library (Sonoma 14.3.1) empty? Based on the report I would assume it should contain 24 images?
2. How can I identify the 28 errors? Using `osxphotos exportdb --errors` generates a file with 1400 issues. Nearly all have the error "Warning: Duplicate Orientation tag in IFD0", is it suggested to manually go through 1400 lines searching for the "other" errors?
`2009/12/13/2009-12-13--img_0662--1746.jpg, E8969663-89A0-4DB3-804E-84849CEC6A84, 2024-03-11 13:29:35.734, {"error": [], "exiftool_error": [], "exiftool_warning": [["/Volumes/nas/…/2009/12/13/2009-12-13--img_0662--1746.jpg", "Warning: Duplicate Orientation tag in IFD0 - /var/folders/ry/p178jh7j7qx7km2phb94r1c80000gn/T/osxphotos_export_g55qqge5/E8969663-89A0-4DB3-804E-84849CEC6A84_exiftool.jpeg"]]}`
3. How do I fix the "Warning: Duplicate Orientation tag in IFD0" error? Or is it a warning I should ignore? I tried using `-m` on `exiftool`, but that did not help.
4. In the errors list I found a few instances of `Error: Not a valid JPEG (looks more like a PNG)`, so I wonder how to fix that issue? I can't find a way to rename the file in Photos.app, do I need to export the images, delete them from the db, change the extension in the finder, then reimport them into the db?

Here's the command I'm using:

```
./osxphotos export \
    --cleanup --verbose --timestamp \
    --export-by-date --album-keyword --person-keyword \
    --download-missing --preview-if-missing --skip-original-if-edited \
    --only-photos --skip-bursts --skip-live --skip-raw --jpeg-ext jpg \
    --not-shared --not-shared-library \
    --exiftool --exiftool-option '-m' \
    --exiftool-merge-keywords --exiftool-merge-persons --favorite-rating \
    --update --update-errors --fix-orientation \
    --filename="{created.date}--{original_name|lower}--{id}" --touch-file --strip --overwrite \
    --ramdb --checkpoint 300 \
    --add-missing-to-album osxphotos-missing \
    --retry 5 \
    /Volumes/nas/…
```

Thanks for any suggestion. And @RhetTbull: This is an awesome project. I'm very thankful for your work on it.

**Comment by RhetTbull** [comment]
First, I wouldn't use `--fix-orientation` This is for a specific issue with iPhotos. It does no harm in your case but it will cause exiftool to re-run on the exported photos every time so that's very time consuming. You shouldn't need this if you're not using iPhotos.

Second, using `--update-errors` will re-export files with errors each time but in your case exiftool will continue to produce errors so this results in unnecessary churn with each update.

> Why is the album for the missing photos in the Photos library (Sonoma 14.3.1) empty?

I don't know; you'd have to look at the report for the missing photos to see what they are, do they appear in Photos, etc. Are the missing photos also hidden? If so, they won't show in the album. It's also possible that the missing photos are syndicated photos (shared via Messages, etc) that haven't been imported to the library yet. If so, they can't be added to an album. 

You can run `osxphotos debug-dump  --dump photos --missing >missing.txt` and send the output to osxphotos.py@gmail.com and I'll take a look to see if there any clues.

> is it suggested to manually go through 1400 lines searching for the "other" errors?

The time stamp is in the output so you can search by date of most recent export.

You can also use the `osxphotos exportdb --report` command to generate an export report for the last report. See `osxphotos help exportdb`. The report will list the errors for the last export among other details.

> How do I fix the "Warning: Duplicate Orientation tag in IFD0" error?

This may be due to use of `--fix-orientation ` See above. But if not, it's an exiftool error and there's nothing osxphotos can do about it.

> In the errors list I found a few instances of `Error: Not a valid JPEG (looks more like a PNG)`, so I wonder how to fix that issue?

This sometimes happens with photos imported from or edited in a 3rd party app. The only solution is to export the image, fix it, the re-import as a new image and delete the old one. This is not an osxphotos issue -- the warning is coming from exiftool. 

> do I need to export the images, delete them from the db, change the extension in the finder, then reimport them into the db?

Yes.

Finally, you don't need `--overwrite` and if run without `--update` this may have undesired consequences (overwrite a previously exported image with a different image).  Like many of the osxphotos options this is one of those that you only need if you know you need it. :-)

**Comment by maia** [comment]
Rhet, thanks for your quick and detailed reply.

> First, I wouldn't use `--fix-orientation` This is for a specific issue with iPhotos. It does no harm in your case but it will cause exiftool to re-run on the exported photos every time so that's very time consuming. You shouldn't need this if you're not using iPhotos.

Okay, will remove it. I was hoping it would fix the "Duplicate Orientation tag" issues.

> Second, using `--update-errors` will re-export files with errors each time but in your case exiftool will continue to produce errors so this results in unnecessary churn with each update.

Oh, I thought it is suggested to use this option to ensure that resolved errors are backuped properly?

> I don't know; you'd have to look at the report for the missing photos to see what they are, do they appear in Photos, etc. Are the missing photos also hidden? If so, they won't show in the album. It's also possible that the missing photos are syndicated photos (shared via Messages, etc) that haven't been imported to the library yet. If so, they can't be added to an album.
> 
> You can run `osxphotos debug-dump --dump photos --missing >missing.txt` and send the output to [osxphotos.py@gmail.com](mailto:osxphotos.py@gmail.com) and I'll take a look to see if there any clues.

Thanks. Looking through the generated (30k lines) report it seems many have `'source': 'nachrichten'` (=Messages) and `'syndicated': True` and `'saved_to_library': False`. I don't know how to show all syndicated photos in Photos.app, so I can delete them, so I now went through the list of filenames manually. Others have filenames ending in `.MP4`, so I wonder why they were even considered when using `--only-photos`. When searching for these filenames in Photos.app I have zero results, so I don't know what's the story with these mp4 files.

> > is it suggested to manually go through 1400 lines searching for the "other" errors?
> 
> The time stamp is in the output so you can search by date of most recent export.

All 1400 lines are from the last export. I ran two exports today, the first crapped out 1 image before completion, meaning no `.osxphotos_export.db` was written, so these are all from one export. I assume I have 1360 or so photos with "Warning: Duplicate Orientation tag in IFD0" error (that's 8% of all photos in the db!) and the other ~40 are the missing photos mentioned above?
 
> You can also use the `osxphotos exportdb --report` command to generate an export report for the last report. See `osxphotos help exportdb`. The report will list the errors for the last export among other details.

I've seen that. But my issue is the sheer amount of errors that might not be errors but just warnings, as the return string has `error: 28`, so it's hard to find those photos that are really errors. It's not ideal if you have to read through 1400 lines of report identifying the few times of real errors, so I was hoping for a different way.
 
> This may be due to use of `--fix-orientation ` See above. But if not, it's an exiftool error and there's nothing osxphotos can do about it.

It's not, as I used `--fix-orientation` hoping it would fix the issue. Is there a way to prevent being spammed with warnings in osxphotos? Are these 8% photos of my database properly saved or not, will they be re-saved on each subsequent run? I'm not sure how to deal with that issue. I can't find any related info on the exifinfo pages.

> This sometimes happens with photos imported from or edited in a 3rd party app. The only solution is to export the image, fix it, the re-import as a new image and delete the old one. This is not an osxphotos issue -- the warning is coming from exiftool.

Thanks. I will have to deal with these manually.

> Finally, you don't need `--overwrite` and if run without `--update` this may have undesired consequences (overwrite a previously exported image with a different image). Like many of the osxphotos options this is one of those that you only need if you know you need it. :-)

Today I realised that I had tons of duplicates saved on the NAS, all with filenames with suffixes (1) and (2). Maybe `--cleanup` would have cleaned up the mess, but I decided to start from scratch and create unique filenames to prevent duplicates. I fear that omitting `--cleanup` might create duplicates again at some point in the future?

**Reply by RhetTbull** [reply]
> Oh, I thought it is suggested to use this option to ensure that resolved errors are backuped properly?

If exiftool produces an error, the file is still backed up. It's just that the metadata may not be embedded properly. There's nothing osxphotos can do about that. There are other causes of errors -- LAN got disconnected, etc -- and the `--update-errors` flag lets you rerun the export for those errors. In your case there's no benefit as exiftool will gladly spit out the errors again.

> Thanks. Looking through the generated (30k lines) report it seems many have `'source': 'nachrichten'` (=Messages) and `'syndicated': True` and `'saved_to_library': False`.

Ok, that's why. These cannot be saved to albums because they aren't really in the library until you save them. I don't think you can delete them. It's a feature of Sonoma. I could modify osxphotos to ignore these upon export. I need to think more about this as the default for osxphotos has always been "try to export everything"

 > I assume I have 1360 or so photos with "Warning: Duplicate Orientation tag in IFD0" error (that's 8% of all photos in the db!) and the other ~40 are the missing photos mentioned above?

You could count the lines:

`grep IFD0 report.csv | wc -l`

Missing photos won't produce errors because there was nothing to edport. The `missing` column in the report will contain a 1.

> It's not ideal if you have to read through 1400 lines of report identifying the few times of real errors,

thats what tools like `grep` are for. OSXPhotos would have no way of knowing which errors are "important". It logs the errors and writes them to the report so you can make that determination. If you use `--update` and avoid rewriting the exiftool data each time you wont see the errors on subsequent exports unless some
metadata changed that required a rewrite. 

> Is there a way to prevent being spammed with warnings in osxphotos?

you could send error output to /dev/null:  `osxphotos export ... 2>/dev/null`. As annoying as these might be osxphotos cannot know which errors are important and so i think it's important to output all errors so the user understands what's going on. 

> Today I realised that I had tons of duplicates saved on the NAS, all with filenames with suffixes (1) and (2). Maybe `--cleanup` would have cleaned up the mess, but I decided to start from scratch and create unique filenames to prevent duplicates. I fear that omitting `--cleanup` might create duplicates again at some point in the future?

the duplicate names only happen when the filename has a collision with exiting filename in the output
directory. OSXPhotos deals with these the exact same way Photos does by appending `(1)` and so on. if you don't want this then use unique filenames as you've done. OSXPhotos keeps track of which file maps to which Photos asset. `--cleanup` has no effect with these if they are files known to osxphotos; it cleans up any files that were not included as part of the backup. That is any file that would have been exported, updated, skipped, or missing, would not be deleted and any other file would be. The primary purpose of this is to delete files from the backup when they are deleted in Photos. For backup, this may or may not be desirable.

**Comment by maia** [comment]
> If exiftool produces an error, the file is still backed up. It's just that the metadata may not be embedded properly. There's nothing osxphotos can do about that. There are other causes of errors -- LAN got disconnected, etc -- and the `--update-errors` flag lets you rerun the export for those errors. In your case there's no benefit as exiftool will gladly spit out the errors again.

Okay, so I should usually not use `--update-errors` unless osxphotos crashes, the lan disconnects etc? I understand.

> Ok, that's why. These cannot be saved to albums because they aren't really in the library until you save them. I don't think you can delete them. It's a feature of Sonoma. I could modify osxphotos to ignore these upon export. I need to think more about this as the default for osxphotos has always been "try to export everything"

Maybe a `--skip-syndicated` option?

> > I assume I have 1360 or so photos with "Warning: Duplicate Orientation tag in IFD0" error (that's 8% of all photos in the db!) and the other ~40 are the missing photos mentioned above?
> 
> You could count the lines:
> 
> `grep IFD0 report.csv | wc -l`

Yeah, 1375 lines. That's 8% of all photos.

> thats what tools like `grep` are for. OSXPhotos would have no way of knowing which errors are "important". 

I will look into the documentation for grep to see how to skip all lines containing a specific string, thanks!

> It logs the errors and writes them to the report so you can make that determination. If you use `--update` and avoid rewriting the exiftool data each time you wont see the errors on subsequent exports unless some metadata changed that required a rewrite.

Are you suggesting not to use `--exiftool` in subsequent runs? I plan to run the script every 1-2 months to backup the photos to the nas.

> you could send error output to /dev/null: `osxphotos export ... 2>/dev/null`. As annoying as these might be osxphotos cannot know which errors are important and so i think it's important to output all errors so the user understands what's going on.

I understand. The confusing part is that osxphotos logs 28 errors but the errors list contains 1400 errors, most of which aren't errors by exiftool definition, they all are `exiftool_warning`. And osxphotos knows that as it wouldn't output 28 errors but 1400 errors, so it's just `exportdb --errors` not being able to differ between "real" errors and warnings.

> > Today I realised that I had tons of duplicates saved on the NAS, all with filenames with suffixes (1) and (2). Maybe `--cleanup` would have cleaned up the mess, but I decided to start from scratch and create unique filenames to prevent duplicates. I fear that omitting `--cleanup` might create duplicates again at some point in the future?
> 
> the duplicate names only happen when the filename has a collision with exiting filename in the output directory. 

Yeah, it duplicated a ton of photos. I was using `--export-by-date` and checked manually, there were tons of duplicates with (1) in the filename which were exactly the same photo as the file without (1). Some runs of osxphotos probably went bad. If `--cleanup` doesn't solve that issue I think I need to use `--overwrite` to prevent file duplication happening in the future.

**Reply by RhetTbull** [reply]
> so it's just `exportdb --errors` not being able to differ between "real" errors and warnings.

`exportdb` is an advanced feature for working with the export database. I do not intend to filter warnings. If you use the `--report` option during export you can easily filter warnings or errors.

> there were tons of duplicates with (1) in the filename which were exactly the same photo as the file without (1).

sounds like you ran `osxphotos export` without using `--update`. In that case, if you do not use `--overwrite` osxphotos will append the suffix to avoid overwriting. It is **not** recommended to export to the same directory without using `--update`

> Are you suggesting not to use `--exiftool` in subsequent runs? I

No. if you use it once you should always use it (or all your files will get updated on next export). OSXPhotos will keep track of the metadata being written to using exiftool and only run exiftool if required.
